From c2de45cbdaee1a8c81b2d65f6bced1439a738e00 Mon Sep 17 00:00:00 2001
From: Daniel Kolesa <daniel@octaforge.org>
Date: Mon, 20 Dec 2021 01:48:49 +0100
Subject: [PATCH 07/15] link stage 2 tools dynamically to libstd

stage 2 tools are the ones that are distributed, and we distribute
them together with the rust libs so there is no need to worry
about unstable ABI

diff --git a/src/bootstrap/src/core/builder/cargo.rs b/src/bootstrap/src/core/builder/cargo.rs
index 0688a1d..e09164d 100644
--- a/src/bootstrap/src/core/builder/cargo.rs
+++ b/src/bootstrap/src/core/builder/cargo.rs
@@ -1148,8 +1148,9 @@ impl Builder<'_> {
         // linking all deps statically into the dylib.
         if matches!(mode, Mode::Std) {
             rustflags.arg("-Cprefer-dynamic");
-        }
-        if matches!(mode, Mode::Rustc) && !self.link_std_into_rustc_driver(target) {
+        } else if matches!(mode, Mode::Rustc) && !self.link_std_into_rustc_driver(target) {
+            rustflags.arg("-Cprefer-dynamic");
+        } else if stage >= 2 {
             rustflags.arg("-Cprefer-dynamic");
         }
 
